File: src/cmcandy/java_language_answers/_0145.java
package cmcandy.java_language_answers;

import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

/**
 * Created by CHEN on 2020/6/3.
 */
public class _0145 {

    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     *     int val;
     *     TreeNode left;
     *     TreeNode right;
     *     TreeNode(int x) { val = x; }
     * }
     */
    static public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }
    class Solution {
        private LinkedList<Integer> res = new LinkedList<>( );
        private Stack<TreeNode> stack = new Stack<>( );
        public List<Integer> postorderTraversal(TreeNode root) {
            // //  递归解法
            if (root==null) return res;
            // if(root.left!=null) postorderTraversal(root.left);
            // if(root.right!=null) postorderTraversal(root.right);
            // res.add(root.val);
            // return res;

            //迭代解法
            //正常遍历，就是反着插入就行
            stack.push(root);
            TreeNode cur ;
            while (!stack.isEmpty( )) {
                cur = stack.pop();
                res.addFirst(cur.val);
                if (cur.left != null) {
                    stack.push(cur.left);
                }
                if(cur.right!=null){
                    stack.push(cur.right);
                }
            }
            return res;
        }
    }
}

File: src/cmcandy/C_Language_Answers/_0145.c

#include <stdio.h>
#include <stdlib.h>
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};
void getTreeSize(struct TreeNode *root, int *returnSize)
{
    if (root == NULL)
    {
        (*returnSize) = 0;
        return;
    }
    if (root->left != NULL)
    {
        (*returnSize)++;
        getTreeSize(root->left, returnSize);
    }
    if (root->right != NULL)
    {
        (*returnSize)++;
        getTreeSize(root->right, returnSize);
    }
    return returnSize;
}

int getVal(struct TreeNode *root, int *res, int index)
{
    if (root->left != NULL)
        index = getVal(root->left, res, index);
    if (root->right != NULL)
        index = getVal(root->right, res, index);
    res[index++] = root->val;
    return index;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *postorderTraversal(struct TreeNode *root, int *returnSize)
{
    (*returnSize) = 1;
    getTreeSize(root, returnSize);
    if ((*returnSize) == 0)
    {
        int *res = (int *)malloc(sizeof(int));
        return res;
    }
    int *res = (int *)malloc(sizeof(int) * (*returnSize));

    getVal(root, res, 0);

    return res;
}